### PR TITLE
make block queue into a more generic verification queue and fix block heap size calculation

### PR DESCRIPTION
**Pull request by rphmeier**
in preparation for header-only sync. verification pipelines (currently assumed to be 3-stage) are defined in the `kinds` module.

### CODE REVIEW

**Comment by rphmeier** [comment]
The latest push implements a header queue, but it will give warnings that the `Headers` struct is never used. This will be remedied in the next PR.

**Comment by rphmeier** [comment]
This also fixes calculation of heap size for blocks in the blocks queue (previously hard-coded to be zero, which was wrong)

**Comment by gavofyork** [comment]
looks fairly reasonable to me.

@arkpar could you take a look?

**Comment by rphmeier** [comment]
Doing a full sync with this right now, strangely it seems to have broken the limits on the queue size -- I saw 100MiB of blocks (48k) queued at once with the default maximums: 50MiB and 30k blocks. I haven't changed the `is_full` logic, so I'm not sure why it seems to be importing too many blocks.

**Comment by rphmeier** [comment]
On the plus side, it's definitely queuing way fewer blocks (while syncing ~700k it has typically 15-20k at any given time) while keeping the usual queue size slightly over 50MiB. So the footprint is definitely more in line with the specified options now.

File: ethcore/src/verification/queue/kind.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+				}
+			}
+		}
+
+		fn verify(un: Self::Unverified, engine: &Engine) -> Result<Self::Verified, Error> {
+			let hash = un.hash();
+			match verify_block_unordered(un.header, un.bytes, engine) {
+				Ok(verified) => Ok(verified),
+				Err(e) => {
+					warn!(target: "client", "Stage 2 block verification failed for {}\n Error: {:?}", hash, e);
```
Comment (arkpar) [inline ethcore/src/verification/queue/kind.rs]: Word "Error" should be removed from this message. It is confusing for users.

File: ethcore/src/verification/queue/kind.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	impl Kind for Blocks {
+		type Input = Unverified;
+		type Unverified = Unverified;
+		type Verified = PreverifiedBlock;
+
+		fn create(input: Self::Input, engine: &Engine) -> Result<Self::Unverified, Error> {
+			match verify_block_basic(&input.header, &input.bytes, engine) {
+				Ok(()) => Ok(input),
+				Err(e) => {
+					warn!(target: "client", "Stage 1 block verification failed for {:?}", input.hash());
```
Comment (arkpar) [inline ethcore/src/verification/queue/kind.rs]: Please add error info to the message

**Pull request merged**